src/commands/deck.ts

```diff
@@ -0,0 +1,136 @@
+import { ChatInputApplicationCommandData, CommandInteraction, MessageEmbed } from "discord.js";
+import fetch from "node-fetch";
+import { inject, injectable } from "tsyringe";
+import { extractURLs, parseURL, TypedDeck } from "ydke";
+import { Command } from "../Command";
+import { getLogger, Logger } from "../logger";
+import { Metrics } from "../metrics";
+
+interface MultiCard {
+	kid: number;
+	password: number;
+	name_en: string;
+}
+
+@injectable()
+export class DeckCommand extends Command {
+	#logger = getLogger("command:deck");
+
+	constructor(@inject(Metrics) metrics: Metrics) {
+		super(metrics);
+	}
+
+	static override get meta(): ChatInputApplicationCommandData {
+		return {
+			name: "deck",
+			description: "Display a deck list from ydke:// format, exported from a number of deck building programs.",
+			options: [
+				{
+					type: "STRING",
+					name: "deck",
+					description: "The ydke:// URL of the deck you want to view.",
+					required: true
+				}
+			]
+		};
+	}
+
+	static override get aliases(): string[] {
+		return [];
+	}
+
+	protected override get logger(): Logger {
+		return this.#logger;
+	}
+
+	async generateProfile(deck: TypedDeck): Promise<MessageEmbed> {
+		// get names from API
+		const allUniqueNames = new Set([...deck.main, ...deck.extra, ...deck.side]);
+		// TODO: decide if we're making a module for API interaction or using fetch directly in commands
+		const cards: MultiCard[] = await (
+			await fetch(`${process.env.SEARCH_API}/multi?password=${[...allUniqueNames].join(",")}`)
+		).json();
+		// we fetch the name before counting because doing this with an array is easier than a record
+		// as such we memoise to avoid duplicate searches
+		const nameMemo: Record<number, string> = {};
+		const getName = (password: number): string => {
+			if (!(password in nameMemo)) {
+				const result = cards.filter(c => c.password === password);
+				if (result.length > 0) {
+					nameMemo[password] = result[0].name_en;
+				} else {
+					nameMemo[password] = password.toString();
+				}
+			}
+			return nameMemo[password];
+		};
+		const namedDeck = {
+			main: [...deck.main].map(getName),
+			extra: [...deck.extra].map(getName),
+			side: [...deck.side].map(getName)
+		};
+		// count the number of each card in the deck
+		const count = (acc: Record<string, number>, val: string): Record<string, number> => {
+			acc[val] = ++acc[val] || 1;
```

Comment:
Is this how we usually do it? This can be confusing because the increment operator modifies the value already.

Reply:
This was a one-liner I found while trying to remind myself how to do it but I suppose a ternary like `acc[val] ? acc[val]+1 : 1` would be clearer.